### ISSUE TITLE
False positive in -Wthread-safety-analysis when returning scoped lock from function

### ISSUE THREAD

**Issue by aaronpuchert** [llvm/llvm-project#58480]
[D129755](https://reviews.llvm.org/D129755) uncovered issues in our handling of scoped locks at the end of a function, see the added test cases:
```c++
MutexLock lock() EXCLUSIVE_LOCK_FUNCTION(mutex) {
  return MutexLock(&mutex);
} // warning: mutex 'mutex' is still held at the end of function
```

**Comment by aaronpuchert** [comment]
Fixed in #105526.
